### PR TITLE
add NetworkPolicy plugin

### PR DESCRIPTION
**Pull request by danwinship**
This adds a new plugin, `redhat/openshift-ovs-networkpolicy`, which implements NetworkPolicy (and does not implement traditional openshift-sdn multitenancy). This is considered "tech preview" for 3.5 so it doesn't have to be completely flawless (eg, the pod watching code is currently not very scalable; there's a trello card about that).

Still to do:
 - [ ] docs
 - [ ] integrate with existing tests/extended/networking tests
 - [ ] pull in the upstream NetworkPolicy test case

@openshift/networking PTAL

### CODE REVIEW

**Comment by knobunc** [comment]
For reference, the spec is at https://github.com/kubernetes/community/blob/master/contributors/design-proposals/network-policy.md

**Comment by knobunc** [comment]
@dcbw can you find some time to look at this soon please?

**Comment by knobunc** [comment]
[test]

**Comment by danwinship** [comment]
Repushed to make the UPSTREAM commit include the entire corresponding upstream commit, and to remove (incorrect) support for string-valued NetworkPolicyPort values.

Note that currently no tests get run with the networkpolicy plugin, so running tests doesn't actually tell you anything (except that the handful of changes outside of networkpolicy.go didn't break anything).

File: pkg/sdn/plugin/networkpolicy.go [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+
+	changed := !existed || !reflect.DeepEqual(oldNPP.flows, npp.flows)
+	if !changed {
+		glog.V(5).Infof("NetworkPolicy %s/%s is unchanged", policy.Namespace, policy.Name)
+	}
+	return changed
+}
+
+func (np *networkPolicyPlugin) watchNetworkPolicies() {
+	RunEventQueue(np.node.kClient.ExtensionsClient.RESTClient(), NetworkPolicies, func(delta cache.Delta) error {
```
Comment (smarterclayton) [inline pkg/sdn/plugin/networkpolicy.go]: We have said we need to stop using event queue.  Adding more uses of it isn't helping :)

File: pkg/sdn/plugin/networkpolicy.go [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+	go utilwait.Forever(np.watchNamespaces, 0)
+	go utilwait.Forever(np.watchNetworkPolicies, 0)
+	return nil
+}
+
+func (np *networkPolicyPlugin) initNamespaces() error {
+	np.lock.Lock()
+	defer np.lock.Unlock()
+
+	namespaces, err := np.node.kClient.Namespaces().List(kapi.ListOptions{})
```
Comment (smarterclayton) [inline pkg/sdn/plugin/networkpolicy.go]: Is this the only network policy that reads namespaces?

File: pkg/sdn/plugin/networkpolicy.go (hunk 1 above)
Comment (danwinship) [inline pkg/sdn/plugin/networkpolicy.go]: The "EventQueue" used in pkg/sdn/plugin is a wrapper around DeltaFIFO (#10070) not the old pkg/client/cache/eventqueue.go.

File: pkg/sdn/plugin/networkpolicy.go (hunk 2 above)
Comment (danwinship) [inline pkg/sdn/plugin/networkpolicy.go]: Yes. With the multitenant policy, everything it needs to track is on the NetNamespace object, but here we need to combine Namespace, NetNamespace, and NetworkPolicies.

**Comment by danwinship** [comment]
pushed a fixup to get Services at least minimally working

**Comment by dcbw** [comment]
LGTM

**Comment by knobunc** [comment]
Since this is tech preview and you have to opt in by manually creating the network policy objects, this feels okay to merge to get some mileage on it before the testing and docs stabilize.  But we need to get those in ASAP. [merge]

**Comment by knobunc** [comment]
Test change LGTM

**Comment by knobunc** [comment]
[merge]

**Comment by knobunc** [comment]
[test]

**Comment by knobunc** [comment]
`pkg/util/ovs/ovs_test.go:138: not enough arguments in call to New`

That was a pain to dig out...

**Comment by eparis** [comment]
[merge]

**Pull request merged**